### PR TITLE
fix font export

### PR DESCRIPTION
**Pull request by akshay-gupta7**
### Why does this PR exist?

Font-family tokens resolved by the Tokens Studio server (Rails/gRPC path) come through as an array-shaped string — e.g. `'["Arial","Helvetica","sans-serif"]'`. On **Export to Variables**, that string was being written to the Figma variable as the literal `[Arial`, because `convertFontFamilyToFigma` split by `,` and stripped quotes without ever removing the surrounding brackets. Font-weights had the same latent issue via `convertFontWeightToFigma`.

### What does this pull request do?

- **`convertFontFamilyToFigma`** — strips surrounding `[...]` before splitting. For JSON-shaped arrays (`["A","B"]`) tries `JSON.parse` first so quoted family names containing commas (`["Font, Name","Arial"]` → `Font, Name`) survive; falls back to a split for single-quoted / unquoted bracket lists.
- **`convertFontWeightToFigma`** — mirrors the bracket-strip so server-form `'["Bold","Regular"]'` no longer flows through untouched.
- **`setValuesOnVariable`** — STRING case simplified. Normalization now lives in the transform functions (right altitude), so the switch only needs a plain-string branch plus the array branch. Numeric-array coercion via `String()` is kept as a belt-and-suspenders safety net.

### Testing this change

1. Connect a Studio-server project that has a font-family token with multiple fallbacks (or a hardcoded `["Arial","Helvetica","sans-serif"]`-shaped value).
2. Run **Export to Variables**.
3. Verify the Figma variable value is `Arial` (previously `[Arial`).
4. Repeat with a font-weight token whose value is an array-shaped string.
5. Unit tests — `yarn jest --testPathPattern="convertFontFamilyToFigma|fontWeight\.test|setValuesOnVariable"` — 40 tests across 4 suites pass, including new coverage for JSON, single-quoted, and unquoted bracket forms plus the comma-in-quoted-name case.

### Additional Notes (if any)

Addresses Copilot review comments (silent JSON parse, added tests for both transforms) and follow-up ultra-review findings (fontWeights parity gap, dead JSON-string branch in `setValuesOnVariable`, `"[]"` literal write, comma-in-quoted-name).

### CODE REVIEW

**Comment by akshay-gupta7** [comment]
# Code Review: PR-3928

**Effort:** max · **Scope:** `main...pr-3928` (4 files, 171 diff lines)

---

## 1. FONT_WEIGHTS parity gap — same bug still present ⚠️ **CONFIRMED**

**File:** `packages/tokens-studio-for-figma/src/plugin/setValuesOnVariable.ts:254`

The array-shaped-string bug this PR fixes for fontFamilies is **not** fixed for fontWeights.

**Trace for input `token.value = '["Bold","Regular"]'` on a FONT_WEIGHTS STRING variable:**

1. `mapTokensToVariableInfo` calls `transformValue(value, 'fontWeights', ..., true)`
2. `transformValue` → `convertFontWeightToFigma(value, true)` → returns `[value]` (fontWeight.ts:2-3 just wraps the input)
3. `token.value` is now `['["Bold","Regular"]']`
4. `setValuesOnVariable` hits the `Array.isArray && typeof value[0] === 'string'` branch
5. Writes `'["Bold","Regular"]'` **verbatim** to the variable

**Fix:** mirror the bracket-strip in `convertFontWeightToFigma`, or handle it in the pre-transform step.

---

## 2. Dead JSON-string branch + misleading test ⚠️ **CONFIRMED**

**File:** `packages/tokens-studio-for-figma/src/plugin/setValuesOnVariable.ts:234`

The `tryParseJson`/splitter branch inside `setValuesOnVariable` is **unreachable in production for FONT_FAMILIES**:

- `mapTokensToVariableInfo` (`utils/mapTokensToVariableInfo.ts:18`) runs `transformValue` on string values
- `transformValue` routes fontFamilies through `convertFontFamilyToFigma(..., true)`, which already strips brackets and returns a single family
- By the time the STRING case runs, a fontFamilies `token.value` can never still be `'["Arial","Helvetica"]'`

The unit test `'FONT_FAMILIES JSON-array string'` passes only because it calls `setValuesOnVariable` directly, bypassing the real pipeline — **it gives false confidence the fix works end-to-end**.

**Fix:** either delete the FONT_FAMILIES portion of the setValuesOnVariable branch, or (better) route the test through `mapTokensToVariableInfo` so it validates the actual production path.

---

## 3. Empty-array bracket string writes literal `"[]"` ⚠️ **CONFIRMED**

**File:** `packages/tokens-studio-for-figma/src/plugin/setValuesOnVariable.ts:244`

For `token.value = '[]'`:

- `trimmed` starts/ends with brackets → enter parse branch
- `tryParseJson` skipped (no `"`)
- `inner = ''`, `firstRaw = ''`, `firstUnquoted = ''` → length 0
- Fallback: `stringValue = token.value = '[]'`
- Variable ends up with literal `"[]"` as its font family / weight

**Fix:** when the fallback yields empty, set `stringValue = undefined` so the write is skipped entirely.

---

## 4. Numeric FONT_WEIGHTS arrays silently dropped 🟡 **PLAUSIBLE**

**File:** `packages/tokens-studio-for-figma/src/plugin/setValuesOnVariable.ts:246`

**Old code:**
```ts
else if (token.type === TokenTypes.FONT_WEIGHTS && Array.isArray(token.value)) {
  setStringValuesOnVariable(variable, mode, token.value[0], hasMetadataChanged);
}
```

**New code** additionally requires `typeof token.value[0] === 'string'`.

Input `[400, 500]` on a STRING variable now falls through with `stringValue === undefined` and the write is skipped. Previously it would have called `setStringValuesOnVariable(..., 400)`.

Very edge case (numeric weights typically create FLOAT variables), but no test locks the behavior in either direction — worth an explicit decision.

---

## 5. Quoted family names containing a comma get truncated 🟡 **PLAUSIBLE**

**File:** `packages/tokens-studio-for-figma/src/plugin/figmaTransforms/convertFontFamilyToFigma.ts:12`

For input `'["Font, Name","Arial"]'`:

- `withoutBrackets = '"Font, Name","Arial"'`
- `split(',')` → `['"Font', ' Name"', '"Arial"']`
- First element `.trim().replace(/['"]/g,'')` → `'Font'`

Variable ends up with `Font` instead of `Font, Name`. Pre-existing bug for non-bracketed input, but the PR's example (`'["Arial","Helvetica"]'`) proves the server can emit JSON — where quoted-with-comma names are legal.

**Fix:** for bracketed inputs, attempt `JSON.parse` first (mirroring the approach in `setValuesOnVariable`); fall back to split only on parse failure.

---

## Priority

| # | Severity | Action |
|---|---|---|
| 1 | **High** — real user-visible bug on the sibling code path | Fix before merge |
| 2 | Medium — dead code + false-signal test | Fix before merge, or file follow-up |
| 3 | Low — narrow edge case | Fix in this PR (one-liner) |
| 4 | Low — behavior change, undocumented | Decide + add test |
| 5 | Low — pre-existing, but PR's context makes it newly relevant | Follow-up PR OK |

**Comment by akshay-gupta7** [comment]
# Code Review: PR-3928

**Effort:** max · **Scope:** `main...pr-3928` (4 files, 171 diff lines)

---

## 1. FONT_WEIGHTS parity gap — same bug still present ⚠️ **CONFIRMED**

**File:** `packages/tokens-studio-for-figma/src/plugin/setValuesOnVariable.ts:254`

The array-shaped-string bug this PR fixes for fontFamilies is **not** fixed for fontWeights.

**Trace for input `token.value = '["Bold","Regular"]'` on a FONT_WEIGHTS STRING variable:**

1. `mapTokensToVariableInfo` calls `transformValue(value, 'fontWeights', ..., true)`
2. `transformValue` → `convertFontWeightToFigma(value, true)` → returns `[value]` (fontWeight.ts:2-3 just wraps the input)
3. `token.value` is now `['["Bold","Regular"]']`
4. `setValuesOnVariable` hits the `Array.isArray && typeof value[0] === 'string'` branch
5. Writes `'["Bold","Regular"]'` **verbatim** to the variable

**Fix:** mirror the bracket-strip in `convertFontWeightToFigma`, or handle it in the pre-transform step.

---

## 2. Dead JSON-string branch + misleading test ⚠️ **CONFIRMED**

**File:** `packages/tokens-studio-for-figma/src/plugin/setValuesOnVariable.ts:234`

The `tryParseJson`/splitter branch inside `setValuesOnVariable` is **unreachable in production for FONT_FAMILIES**:

- `mapTokensToVariableInfo` (`utils/mapTokensToVariableInfo.ts:18`) runs `transformValue` on string values
- `transformValue` routes fontFamilies through `convertFontFamilyToFigma(..., true)`, which already strips brackets and returns a single family
- By the time the STRING case runs, a fontFamilies `token.value` can never still be `'["Arial","Helvetica"]'`

The unit test `'FONT_FAMILIES JSON-array string'` passes only because it calls `setValuesOnVariable` directly, bypassing the real pipeline — **it gives false confidence the fix works end-to-end**.

**Fix:** either delete the FONT_FAMILIES portion of the setValuesOnVariable branch, or (better) route the test through `mapTokensToVariableInfo` so it validates the actual production path.

---

## 3. Empty-array bracket string writes literal `"[]"` ⚠️ **CONFIRMED**

**File:** `packages/tokens-studio-for-figma/src/plugin/setValuesOnVariable.ts:244`

For `token.value = '[]'`:

- `trimmed` starts/ends with brackets → enter parse branch
- `tryParseJson` skipped (no `"`)
- `inner = ''`, `firstRaw = ''`, `firstUnquoted = ''` → length 0
- Fallback: `stringValue = token.value = '[]'`
- Variable ends up with literal `"[]"` as its font family / weight

**Fix:** when the fallback yields empty, set `stringValue = undefined` so the write is skipped entirely.

---

## 4. Numeric FONT_WEIGHTS arrays silently dropped 🟡 **PLAUSIBLE**

**File:** `packages/tokens-studio-for-figma/src/plugin/setValuesOnVariable.ts:246`

**Old code:**
```ts
else if (token.type === TokenTypes.FONT_WEIGHTS && Array.isArray(token.value)) {
  setStringValuesOnVariable(variable, mode, token.value[0], hasMetadataChanged);
}
```

**New code** additionally requires `typeof token.value[0] === 'string'`.

Input `[400, 500]` on a STRING variable now falls through with `stringValue === undefined` and the write is skipped. Previously it would have called `setStringValuesOnVariable(..., 400)`.

Very edge case (numeric weights typically create FLOAT variables), but no test locks the behavior in either direction — worth an explicit decision.

---

## 5. Quoted family names containing a comma get truncated 🟡 **PLAUSIBLE**

**File:** `packages/tokens-studio-for-figma/src/plugin/figmaTransforms/convertFontFamilyToFigma.ts:12`

For input `'["Font, Name","Arial"]'`:

- `withoutBrackets = '"Font, Name","Arial"'`
- `split(',')` → `['"Font', ' Name"', '"Arial"']`
- First element `.trim().replace(/['"]/g,'')` → `'Font'`

Variable ends up with `Font` instead of `Font, Name`. Pre-existing bug for non-bracketed input, but the PR's example (`'["Arial","Helvetica"]'`) proves the server can emit JSON — where quoted-with-comma names are legal.

**Fix:** for bracketed inputs, attempt `JSON.parse` first (mirroring the approach in `setValuesOnVariable`); fall back to split only on parse failure.

---

## Priority

| # | Severity | Action |
|---|---|---|
| 1 | **High** — real user-visible bug on the sibling code path | Fix before merge |
| 2 | Medium — dead code + false-signal test | Fix before merge, or file follow-up |
| 3 | Low — narrow edge case | Fix in this PR (one-liner) |
| 4 | Low — behavior change, undocumented | Decide + add test |
| 5 | Low — pre-existing, but PR's context makes it newly relevant | Follow-up PR OK |

File: packages/tokens-studio-for-figma/src/plugin/figmaTransforms/convertFontFamilyToFigma.ts [hunk 1]
```diff
@@ -2,7 +2,14 @@ export function convertFontFamilyToFigma(value: string, shouldOutputForVariables
...
     if (shouldOutputForVariables) {
-      const fontFamilies = stringValue.split(',');
+      // Studio's server resolver returns fontFamilies as an array-shaped string
+      // (e.g. '["Arial","Helvetica"]'). Strip surrounding brackets before splitting
+      // so we don't end up with '[Arial' as the "first" family.
+      const trimmed = stringValue.trim();
+      const withoutBrackets = trimmed.startsWith('[') && trimmed.endsWith(']')
+        ? trimmed.slice(1, -1)
+        : trimmed;
+      const fontFamilies = withoutBrackets.split(',');
```
[the remark this answers is not in the document]
Comment (akshay-gupta7) [reply packages/tokens-studio-for-figma/src/plugin/figmaTransforms/convertFontFamilyToFigma.ts]: resolved in https://github.com/tokens-studio/figma-plugin/pull/3928/commits/a0e3ed5d9c1324e0a1422c96b552e69c6f9523e8

File: packages/tokens-studio-for-figma/src/plugin/setValuesOnVariable.ts [hunk 2]
```diff
@@ -11,6 +11,7 @@ import { transformValue } from './helpers';
 import { variableWorker } from './Worker';
 import { ProgressTracker } from './ProgressTracker';
 import { checkVariableAliasEquality } from '@/utils/checkVariableAliasEquality';
+import { tryParseJson } from '@/utils/tryParseJson';
```
[the remark this answers is not in the document]
Comment (akshay-gupta7) [reply packages/tokens-studio-for-figma/src/plugin/setValuesOnVariable.ts]: resolved in https://github.com/tokens-studio/figma-plugin/pull/3928/commits/a0e3ed5d9c1324e0a1422c96b552e69c6f9523e8

File: packages/tokens-studio-for-figma/src/plugin/setValuesOnVariable.ts [hunk 3]
```diff
@@ -223,15 +224,45 @@ export default async function setValuesOnVariable(
...
-                  setStringValuesOnVariable(variable, mode, token.value, hasMetadataChanged);
+                  // Studio's server-resolved fontFamilies/fontWeights arrive as an array-shaped string.
+                  // Usually JSON (e.g. '["Arial","Helvetica"]') but can also be single-quoted or unquoted
+                  // (e.g. "['Arial', 'Helvetica']" or "[Arial, Helvetica, sans-serif]"). Figma variables
+                  // hold a single string, so extract and take the first entry.
+                  const trimmed = token.value.trim();
+                  if (isFontMultiValueType && trimmed.startsWith('[') && trimmed.endsWith(']')) {
+                    const parsed = tryParseJson<unknown[]>(trimmed);
+                    if (Array.isArray(parsed) && typeof parsed[0] === 'string') {
+                      stringValue = parsed[0].trim();
```
[the remark this answers is not in the document]
Comment (akshay-gupta7) [reply packages/tokens-studio-for-figma/src/plugin/setValuesOnVariable.ts]: resolved in https://github.com/tokens-studio/figma-plugin/pull/3928/commits/a0e3ed5d9c1324e0a1422c96b552e69c6f9523e8

File: packages/tokens-studio-for-figma/src/plugin/setValuesOnVariable.ts [hunk 4]
```diff
@@ -223,15 +224,45 @@ export default async function setValuesOnVariable(
...
+
                 if (typeof token.value === 'string' && !token.value.includes('{')) {
-                  setStringValuesOnVariable(variable, mode, token.value, hasMetadataChanged);
+                  // Studio's server-resolved fontFamilies/fontWeights arrive as an array-shaped string.
+                  // Usually JSON (e.g. '["Arial","Helvetica"]') but can also be single-quoted or unquoted
+                  // (e.g. "['Arial', 'Helvetica']" or "[Arial, Helvetica, sans-serif]"). Figma variables
+                  // hold a single string, so extract and take the first entry.
+                  const trimmed = token.value.trim();
+                  if (isFontMultiValueType && trimmed.startsWith('[') && trimmed.endsWith(']')) {
+                    const parsed = tryParseJson<unknown[]>(trimmed);
```
[the remark this answers is not in the document]
Comment (akshay-gupta7) [reply packages/tokens-studio-for-figma/src/plugin/setValuesOnVariable.ts]: resolved in https://github.com/tokens-studio/figma-plugin/pull/3928/commits/a0e3ed5d9c1324e0a1422c96b552e69c6f9523e8

**Comment by anathaniel-TS** [comment]
🛡️ **Hyma Compliance Check**

✅ **No compliance-relevant changes detected**

7 file(s) scanned · 0 findings

> ℹ️ This check is informational.
> 🛡️ Hyma Compliance Check · automated

**Comment by anathaniel-TS** [comment]
🛡️ **Hyma Compliance Check**

✅ **No compliance-relevant changes detected**

3 file(s) scanned · 0 findings

> ℹ️ This check is informational.
> 🛡️ Hyma Compliance Check · automated

**Comment by anathaniel-TS** [comment]
🛡️ **Hyma Compliance Check**

✅ **No compliance-relevant changes detected**

7 file(s) scanned · 0 findings

> ℹ️ This check is informational.
> 🛡️ Hyma Compliance Check · automated

**Comment by anathaniel-TS** [comment]
🛡️ **Hyma Compliance Check**

✅ **No compliance-relevant changes detected**

7 file(s) scanned · 0 findings

> ℹ️ This check is informational.
> 🛡️ Hyma Compliance Check · automated

**Comment by anathaniel-TS** [comment]
🛡️ **Hyma Compliance Check**

✅ **No compliance-relevant changes detected**

3 file(s) scanned · 0 findings

> ℹ️ This check is informational.
> 🛡️ Hyma Compliance Check · automated

**Comment by anathaniel-TS** [comment]
🛡️ **Hyma Compliance Check**

✅ **No compliance-relevant changes detected**

6 file(s) scanned · 0 findings

> ℹ️ This check is informational.
> 🛡️ Hyma Compliance Check · automated

**Comment by anathaniel-TS** [comment]
🛡️ **Hyma Compliance Check**

✅ **No compliance-relevant changes detected**

6 file(s) scanned · 0 findings

> ℹ️ This check is informational.
> 🛡️ Hyma Compliance Check · automated

File: packages/tokens-studio-for-figma/src/plugin/figmaTransforms/fontWeight.ts [hunk 5]
```diff
@@ -1,5 +1,14 @@
 export function convertFontWeightToFigma(value: string, shouldOutputForVariables = false): string[] {
   if (shouldOutputForVariables) {
+    // Studio's server resolver returns fontWeights as an array-shaped string
+    // (e.g. '["Bold","Regular"]'). Strip surrounding brackets and take the first
+    // entry so we don't end up writing the raw JSON literal to the variable.
+    const trimmed = value.trim();
```
Comment (six7) [inline packages/tokens-studio-for-figma/src/plugin/figmaTransforms/fontWeight.ts]: why add array-based support to font weights? this isnt even possible in Studio

Comment (akshay-gupta7) [reply packages/tokens-studio-for-figma/src/plugin/figmaTransforms/fontWeight.ts]: addressed in https://github.com/tokens-studio/figma-plugin/pull/3928/commits/ce633500f1710bfef15894735a54b03feaf25bbd

**Comment by anathaniel-TS** [comment]
🛡️ **Hyma Compliance Check**

✅ **No compliance-relevant changes detected**

1 file(s) scanned · 0 findings

> ℹ️ This check is informational.
> 🛡️ Hyma Compliance Check · automated

**Comment by anathaniel-TS** [comment]
🛡️ **Hyma Compliance Check**

✅ **No compliance-relevant changes detected**

3 file(s) scanned · 0 findings

> ℹ️ This check is informational.
> 🛡️ Hyma Compliance Check · automated

**Review by six7 — APPROVED** [review]
